File: apps/web/src/types/bid-package.ts
export interface BidSubmission {
  id: string;
  bid_package_id: string;
  sub_company_id: string;
  sub_company_name?: string | null;
  total_amount_cents: number;
  line_items: Record<string, unknown>[];
  qualifications?: string | null;
  schedule_duration_days?: number | null;
  exclusions?: string | null;
  inclusions?: string | null;
  notes?: string | null;
  status: string;
  submitted_at?: string | null;
  created_at: string;
}

export interface BidPackage {
  id: string;
  project_id: string;
  number: number;
  formatted_number: string;
  title: string;
  description?: string | null;
  trade?: string | null;
  trades: string[];
  status: string;
  bid_due_date?: string | null;
  pre_bid_meeting_date?: string | null;
  estimated_value_cents: number;
  requirements?: string | null;
  scope_documents: unknown[];
  invited_sub_ids: string[];
  submission_count: number;
  awarded_sub_id?: string | null;
  awarded_at?: string | null;
  created_by?: string | null;
  created_by_name?: string | null;
  created_at: string;
  updated_at: string;
}

export interface BidPackageCreateInput {
  title: string;
  description?: string;
  trade?: string;
  trades?: string[];
  bid_due_date?: string;
  estimated_value_cents?: number;
  requirements?: string;
}

export interface BidPackageUpdateInput {
  title?: string;
  description?: string;
  trade?: string;
  trades?: string[];
  bid_due_date?: string;
  estimated_value_cents?: number;
  requirements?: string;
}

export interface BidSubmissionCreateInput {
  total_amount_cents?: number;
  line_items?: Record<string, unknown>[];
  qualifications?: string;
  schedule_duration_days?: number;
  exclusions?: string;
  inclusions?: string;
  notes?: string;
}

export interface BidComparisonResponse {
  submissions: BidSubmission[];
  lowest_amount_cents: number;
  highest_amount_cents: number;
  average_amount_cents: number;
  recommended_submission_id?: string | null;
}

File: apps/api/app/schemas/bid_package.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.common import PaginationMeta


class BidPackageCreate(BaseModel):
    title: str
    description: Optional[str] = None
    trade: Optional[str] = None
    trades: list[str] = []
    bid_due_date: Optional[datetime] = None
    pre_bid_meeting_date: Optional[datetime] = None
    estimated_value_cents: Optional[int] = None
    requirements: Optional[str] = None
    scope_documents: list = []


class BidPackageUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    trade: Optional[str] = None
    trades: Optional[list[str]] = None
    bid_due_date: Optional[datetime] = None
    pre_bid_meeting_date: Optional[datetime] = None
    estimated_value_cents: Optional[int] = None
    requirements: Optional[str] = None
    scope_documents: Optional[list] = None


class DistributeBidPackageRequest(BaseModel):
    invited_sub_ids: list[UUID]


class AwardBidRequest(BaseModel):
    submission_id: UUID
    trade: Optional[str] = None


class BidSubmissionCreate(BaseModel):
    total_amount_cents: Optional[int] = None
    line_items: list = []
    qualifications: Optional[str] = None
    schedule_duration_days: Optional[int] = None
    exclusions: Optional[str] = None
    inclusions: Optional[str] = None
    notes: Optional[str] = None


class BidSubmissionUpdate(BaseModel):
    total_amount_cents: Optional[int] = None
    line_items: Optional[list] = None
    qualifications: Optional[str] = None
    schedule_duration_days: Optional[int] = None
    exclusions: Optional[str] = None
    inclusions: Optional[str] = None
    notes: Optional[str] = None


class BidSubmissionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    bid_package_id: UUID
    sub_company_id: UUID
    sub_company_name: Optional[str] = None
    total_amount_cents: int = 0
    line_items: list = []
    qualifications: Optional[str] = None
    schedule_duration_days: Optional[int] = None
    exclusions: Optional[str] = None
    inclusions: Optional[str] = None
    notes: Optional[str] = None
    status: str
    submitted_at: Optional[datetime] = None
    created_at: datetime


class BidComparisonResponse(BaseModel):
    submissions: list[BidSubmissionResponse]
    lowest_amount_cents: int = 0
    highest_amount_cents: int = 0
    average_amount_cents: int = 0
    recommended_submission_id: Optional[UUID] = None


class BidPackageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    number: int
    formatted_number: str
    title: str
    description: Optional[str] = None
    trade: Optional[str] = None
    trades: list[str] = []
    status: str
    bid_due_date: Optional[datetime] = None
    pre_bid_meeting_date: Optional[datetime] = None
    estimated_value_cents: int = 0
    requirements: Optional[str] = None
    scope_documents: list = []
    invited_sub_ids: list = []
    submission_count: int = 0
    awarded_sub_id: Optional[UUID] = None
    awarded_at: Optional[datetime] = None
    created_by: Optional[UUID] = None
    created_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class BidPackageListResponse(BaseModel):
    data: list[BidPackageResponse]
    meta: PaginationMeta
